ledger/src/blockstore.rs

```diff
@@ -1307,108 +1359,69 @@ impl Blockstore {
-    // indexes in the ledger in the range [start_index, end_index)
-    // for the slot with the specified slot
-    fn find_missing_indexes<C>(
-        db_iterator: &mut DBRawIterator,
+    pub fn find_missing_data_indexes(
+        &self,
         slot: Slot,
-        first_timestamp: u64,
         start_index: u64,
         end_index: u64,
         max_missing: usize,
-    ) -> Vec<u64>
-    where
-        C: Column<Index = (u64, u64)>,
-    {
-        if start_index >= end_index || max_missing == 0 {
-            return vec![];
-        }
-
-        let mut missing_indexes = vec![];
-        let ticks_since_first_insert =
-            DEFAULT_TICKS_PER_SECOND * (timestamp() - first_timestamp) / 1000;
-
-        // Seek to the first shred with index >= start_index
-        db_iterator.seek(&C::key((slot, start_index)));
-
-        // The index of the first missing shred in the slot
-        let mut prev_index = start_index;
-        'outer: loop {
-            if !db_iterator.valid() {
-                for i in prev_index..end_index {
-                    missing_indexes.push(i);
-                    if missing_indexes.len() == max_missing {
-                        break;
+    ) -> Vec<u64> {
+        self.find_missing_data_indexes_ts(slot, 0, start_index, end_index, max_missing)
+    }
+    fn find_data_indexes(
+        &self,
+        slot: Slot,
+        start_index: u64,
+        end_index: u64,
+        first_ts: &mut u64,
+    ) -> HashSet<u64> {
+        let dir = fs::read_dir(self.slot_data_dir(slot));
+        let min_ts = *first_ts;
+        if let Ok(dir) = dir {
+            dir.filter_map(|e| {
+                let e = e.ok()?;
+                let ix: u64 = std::str::FromStr::from_str(e.file_name().to_str()?).ok()?;
+                if ix >= start_index || ix <= end_index {
+                    if min_ts > 0 {
+                        let ts = fs::metadata(e.path())
+                            .ok()?
+                            .modified()
+                            .ok()?
+                            .duration_since(std::time::UNIX_EPOCH)
+                            .ok()?;
+                        let ts = duration_as_ms(&ts);
+                        if ts > min_ts {
+                            *first_ts = cmp::min(ts, *first_ts);
+                        }
                     }
-                }
-                break;
-            }
-            let (current_slot, index) = C::index(&db_iterator.key().expect("Expect a valid key"));
-
-            let current_index = {
-                if current_slot > slot {
-                    end_index
+                    Some(ix)
                 } else {
-                    index
-                }
-            };
-
-            let upper_index = cmp::min(current_index, end_index);
-            // the tick that will be used to figure out the timeout for this hole
-            let reference_tick = u64::from(Shred::reference_tick_from_data(
```

Comment:
This can be done in a separate PR, but we can remove all the related `reference_tick_from_data` logic now. We've simplified the logic to just not repairing for the first half of a slot after receiving the first shred, right?